myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/TaskFactory.java

```diff
@@ -125,82 +109,40 @@ private String getConfigurationUrl() {
-        //The url for the resource manager config is: http(s)://hostname:port/conf so fetcher.cpp downloads the
-        //config file to conf, It's an xml file with the parameters of yarn-site.xml, core-site.xml and hdfs.xml.
-        String configCopyCmd = "cp conf " + cfg.getYarnEnvironment().get("YARN_HOME") +
-            "/etc/hadoop/yarn-site.xml";
-
-        //Command to run the executor
-        String executorPathString = myriadExecutorConfiguration.getPath();
-        String executorCmd = "export CAPSULE_CACHE_DIR=`pwd`;echo $CAPSULE_CACHE_DIR; " +
-            "sudo -E -u " + cfg.getFrameworkUser().get() + " -H " +
-            "java -Dcapsule.log=verbose -jar " + getFileName(executorPathString);
-
-        //Concatenate all the subcommands
-        String cmd = tarCmd + "&&" + chownCmd + "&&" + configCopyCmd + "&&" + executorCmd;
+        String nodeManagerUri = myriadExecutorConfiguration.getNodeManagerUri().get();
+        clGenerator = new DownloadNMExecutorCLGenImpl(cfg, profile, ports, nodeManagerUri);
+        cmd = clGenerator.generateCommandLine();
 
         //get the nodemanagerURI
         //We're going to extract ourselves, so setExtract is false
-        LOGGER.info("Getting Hadoop distribution from:" + nmURIString);
-        URI nmUri = URI.newBuilder().setValue(nmURIString).setExtract(false)
-            .build();
+        LOGGER.info("Getting Hadoop distribution from:" + nodeManagerUri);
+        URI nmUri = URI.newBuilder().setValue(nodeManagerUri).setExtract(false).build();
 
         //get configs directly from resource manager
         String configUrlString = getConfigurationUrl();
         LOGGER.info("Getting config from:" + configUrlString);
         URI configUri = URI.newBuilder().setValue(configUrlString)
-            .build();
-
-        //get the executor URI
-        LOGGER.info("Getting executor from:" + executorPathString);
-        URI executorUri = URI.newBuilder().setValue(executorPathString).setExecutable(true)
-            .build();
-
+          .build();
         LOGGER.info("Slave will execute command:" + cmd);
-        commandInfo.addUris(nmUri).addUris(configUri).addUris(executorUri).setValue("echo \"" + cmd + "\";" + cmd);
-
+        commandInfo.addUris(nmUri).addUris(configUri).setValue("echo \"" + cmd + "\";" + cmd);
         commandInfo.setUser(cfg.getFrameworkSuperUser().get());
 
       } else {
-        String cmdPrefix = "export CAPSULE_CACHE_DIR=`pwd` ;" +
-            "echo $CAPSULE_CACHE_DIR; java -Dcapsule.log=verbose -jar ";
-        String executorPath = myriadExecutorConfiguration.getPath();
-        String cmd = cmdPrefix + getFileName(executorPath);
-        URI executorURI = URI.newBuilder().setValue(executorPath)
-            .setExecutable(true).build();
-        commandInfo.addUris(executorURI)
-            .setValue("echo \"" + cmd + "\";" + cmd);
+        clGenerator = new NMExecutorCLGenImpl(cfg, profile, ports);
```

Comment:
Since the clGenerator that's being chosen is based on myriad config params, the construction of the implementation objects can actually be moved into the Guice module itself. For e.g, in MyriadModule.java, you can define something like below

```java
    @Provides
    @Singleton
    ExecutorCommandLineGenerator providesCLIGenerator(MyriadConfiguration cfg) {
        ExecutorCommandLineGenerator cliGenerator = null;
        if (myriadExecutorConfiguration.getNodeManagerUri().isPresent()) {
            cliGenerator = new DownloadNMExecutorCLGenImpl(cfg);
        } else {
            cliGenerator = new NMExecutorCLGenImpl(cfg);
        }
        return cliGenerator;
    }
```

I see that currently DownloadNMExecutorCLGenImpl and NMExecutorCLGenImpl have additional dependencies in their constructor such as NMProfile and NMPorts. However, I feel that these should actually be passed into the generateCommandLine() method rather than as args into the constructor. One clear advantage is that we avoid creation of these objects and throwing them away after constructing the command line. The second advantage is separation of concerns: TaskFactory need not care how it got a command line generator.